platform/logging/LoggerFactory.ts

```diff
@@ -0,0 +1,109 @@
+    let context = namespace.join('.');
+    if (this.loggers.has(context)) {
+      return this.loggers.get(context)!;
+    }
+
+    let loggerConfig;
+    if (this.config) {
+      loggerConfig = this.getLoggerConfigByContext(this.config, context);
+    } else {
+      // If we don't have config yet, use `buffered` appender that will store all logged messages in the memory
+      // until the config is ready.
+      loggerConfig = { level: LogLevel.All, appenders: [BUFFERED_APPENDER_KEY] };
+    }
+
+    this.loggers.set(context,
+      new LoggerAdapter(
+        new BaseLogger(
+          namespace,
+          loggerConfig.level,
+          loggerConfig.appenders.map((appenderKey) => this.appenders.get(appenderKey)!)
+        )
+      )
+    );
+
+    return this.loggers.get(context)!;
+  }
+
+  updateConfig(config: LoggingConfig) {
+    // TODO: Should we support `config.isEqual()` to avoid updating the config if it's still the same?
+    this.config = config;
+
+    // Before we dispose appenders, let's check whether we have any `buffered` appender that may need to be
+    // flushed via newly configured appenders.
+    const bufferedLogRecords = this.appenders.has(BUFFERED_APPENDER_KEY) ?
+      [...(<BufferAppender>this.appenders.get(BUFFERED_APPENDER_KEY)!).buffer] :
+      [];
+
+    this.updateAppendersFromConfig(config);
+    this.updateLoggersFromConfig(config);
+
+    // Re-log all buffered log records with newly configured appenders.
+    for (const logRecord of bufferedLogRecords) {
+      this.get(logRecord.context.join('.')).log(logRecord);
+    }
+  }
+
+  close(): void {
+    for (const appender of this.appenders.values()) {
+      appender.close();
+    }
+
+    this.appenders.clear();
+    this.loggers.clear();
+  }
+
+  private updateAppendersFromConfig(config: LoggingConfig) {
+    for (const appender of this.appenders.values()) {
+      appender.close();
+    }
+
+    this.appenders.clear();
+
+    // Initialize new log appenders.
+    for (const [appenderKey, appenderConfig] of config.appenders.entries()) {
+      this.appenders.set(appenderKey, appenderConfig.createAppender());
+    }
+  }
+
+  private updateLoggersFromConfig(config: LoggingConfig) {
+    for (const [loggerKey, loggerAdapter] of this.loggers.entries()) {
+      const loggerConfig = this.getLoggerConfigByContext(config, loggerKey);
+      loggerAdapter.logger = new BaseLogger(
+        // TODO: Do we need to store namespace as an array at all?
+        loggerKey.split('.'),
+        loggerConfig.level,
+        loggerConfig.appenders.map((appenderKey) => this.appenders.get(appenderKey)!)
+      );
+    }
+  }
+
+  private getLoggerConfigByContext(config: LoggingConfig, context: string) {
+    // If we don't have dedicated configuration for specified context, let's use config from `root` context,
+    // that is always should be defined (enforced by configuration schema).
+    return config.loggers.get(config.loggers.has(context) ? context : 'root')!;
```

Comment:
> Maybe we also need the ability to override the log level itself? e.g. for ES: https://www.elastic.co/guide/en/elasticsearch/reference/5.1/misc-cluster.html#cluster-logger.

> That would likely have to be a different config, though

I think once I handle your first comment above it will be possible, like this (`appenders` should become optional though, and if none of the logger defines `appenders`, the ones from `root` should be used, I'll think how to make that easier):
```

logging:
  appenders:
    console:
      kind: console
      pattern: "[{timestamp}][{level}][{context}] {message}"
  loggers:
    server:
      appenders: [server-console]
      level: all
    server.foo:
      level: debug
```